src/lib/utils/track.ts

```diff
@@ -0,0 +1,107 @@
+import * as React from "react"
+import { track as _track, Track as _Track } from "react-tracking"
+
+// tslint:disable-next-line:no-namespace
+export namespace Schema {
+  /**
+   * The global tracking-info keys in Artsy’s schema.
+   */
+  export interface Global {
+    /**
+     * The name of an event.
+     */
+    action: string
+
+    /**
+     * The root container component should specify this as the screen context.
+     */
+    page: string
+  }
+
+  export interface Entity extends Global {
+    /**
+     * The ID of the entity in its database. E.g. the Mongo ID for entities that reside in Gravity.
+     */
+    entity_id: string
+
+    /**
+     * The public slug for this entity.
+     */
+    entity_slug: string
+  }
+}
+
+/**
+ * Use this interface to augment the `track` function with props, state, or custom tracking-info schema.
+ *
+ * @example
+ *
+ *      ```ts
+ *      import { Schema, Track, track as _track } from "src/utils/track"
+ *
+ *
+ *      interface Props {
+ *        artist: {
+ *          id: string
+ *          slug: string
+ *        }
+ *      }
+ *
+ *      interface State {
+ *        following: boolean
+ *      }
+ *
+ *      const track: Track<Props, State, Schema.Entity> = _track
```

Comment:
Little observation: it looks like the order of the interfaces is different in this example from the following:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-tracking/test/react-tracking-with-types-tests.tsx#L20

Is that intended or I'm missing something?

Reply:
Good spot. So that’s because with the generic version from DefinitelyTyped it’s more reasonable to expect schema typings, whereas with this version it already has default schema typings (`Schema.Global`) and so it seemed likely that there will be cases where you want to specify `Props` but don’t need to specify different schema typings.

Reply:
What do you think?

Reply:
I think that makes sense ( though we'll want the same thing inside Reaction so that it's consistent between projects )

Reply:
I see, I get the reasoning now but I think it'd be better to keep the order the same globally. Otherwise, we might get confused in our own codebase if we look at the DefinitelyTyped definitions as a reference.

I personally I prefer the version with `Props` first as it's easier to remember and it follows `Component<Props, State>` but either is fine I think.

Reply:
I agree that having the same signature would simplify the understanding a bit, but I also do think the different signatures make sense in their own contexts. I would suggest we give it a try and see if people actually run into issues first, unless others have a preference as well?

Reply:
@orta Aye, I’ll port this back to Emission once a final decision has been taken here.

Reply:
To summarise:

* The generic react-tracking types in DefinitelyTyped take types in the following order, because without any default types it seems more likely that you’ll specify a schema over props and state:
  1. `Schema`, default `any`
  2. `Props`, default `any`
  3. `State`, default `any`
* In our specific types the order is different, because we have a default schema and it seems more likely that you’ll want to specify props or state _without_ having to specify the default schema types:
  1. `Props`, default `any`
  2. `State`, default `any`
  3. `Schema`, default `Schema.Global`

Reply:
I would add that type parameters aren’t really different than other API parameters, and we change order and provide more specific defaults all the time. I know that this all still requires getting used to reading typings, but once comfortable with it it shouldn’t really read differently than any other API would do. (I mean, ideally :D)